client/services/code_evaluators/PythonPrerequisiteCheckService.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for performing pre-requisite checks on Python code snippets.
+ */
+
+tie.factory('PythonPrerequisiteCheckService', [
+  'PYTHON_STANDARD_LIBRARIES',
+  'PrerequisiteEvalResultObjectFactory',
+  function(
+      PYTHON_STANDARD_LIBRARIES,
+      PrerequisiteEvalResultObjectFactory) {
+
+    const LEFT_TRIM_PATTERN = /^\s+/;
+
+    var checkStarterCodePresent = function(starterCode, code) {
```

Comment:
This is far too complicated / ad hoc. It'll be hard to maintain.

So, just to verify: the point of this prereq check is to ensure that the functions in the starter code still exist within the learner's code, right? If so, could we instead restructure the logic in this function roughly as follows?

```
    expectedTopLevelFunctions = extractTopLevelFunctions(starterCode);
    verifyFunctionsExist(code, expectedTopLevelFunctions);
```